File: src/src_core/fmg/core/img/PolarLightLogoTransformer.java
package fmg.core.img;

import fmg.common.HSV;

/** Transforming of logo palette */
public class PolarLightLogoTransformer implements IModelTransformer {

    @Override
    public void execute(IAnimatedModel model) {
        if (!(model instanceof LogoModel))
            throw new RuntimeException("Illegal usage transformer");

        LogoModel lm = (LogoModel)model;

        if (!lm.isPolarLights())
            return;

        double rotateAngleDelta = 360.0 / lm.getTotalFrames(); // 360° / TotalFrames
        if (!lm.getAnimeDirection())
            rotateAngleDelta = -rotateAngleDelta;
        HSV[] palette = lm.getPalette();
        for (int i = 0; i < palette.length; ++i)
            palette[i].h += rotateAngleDelta;
    }

}

File: src/src_core/fmg/core/img/PolarLightLogoTransformer.cs
using System;
using fmg.common;

namespace fmg.core.img {

    /// <summary> Transforming of logo palette </summary>
    public class PolarLightLogoTransformer : IModelTransformer {

        public void Execute(IAnimatedModel model) {
            if (!(model is LogoModel lm))
                throw new InvalidOperationException("Illegal usage transformer");

            if (!lm.PolarLights)
                return;

            double rotateAngleDelta = 360.0 / lm.TotalFrames; // 360° / TotalFrames
            if (!lm.AnimeDirection)
                rotateAngleDelta = -rotateAngleDelta;
            HSV[] palette = lm.Palette;
            for (int i = 0; i < palette.Length; ++i)
                palette[i].h += rotateAngleDelta;
        }

    }

}
